File: func10.py
# func10

class Sum:
    def __init__(self, x):
        self.x = x
        self.concat = lambda o: Sum(x + o.x)
    
    def __repr__(self):
        return 'Sum({})'.format(self.x)
    
    @staticmethod
    def empty():
        return Sum(0)

File: func10.js
// func10

const { Map, List } = require('immutable-ext')

const Sum = x => ({
    x,
    concat: ({x: y}) => Sum(x + y),
    inspect: () => `Sum(${x})`
})

Sum.empty = () => Sum(0)

// const result = List.of(Sum(1), Sum(2), Sum(3))
//                 .fold(Sum.empty())


// const result = Map({brian: 3, sara: 5}).map(Sum).fold(Sum.empty())
const result = Map({brian: 3, sara: 5}).foldMap(Sum, Sum.empty())

console.log(result)